File: cli/node/src/application/onboarding/key.ts
import fs from 'node:fs';
import { identifyAgent, pingApiKey } from '../../backend/auth.js';
import { CONFIG_FILE, loadConfig, saveConfig } from '../../config/store.js';
import { warnOptional } from '../../runtime/warning.js';

/**
 * 探测密钥是否仍可复用；网络故障保留现有账号并告警。
 * @param apiKey - 访问服务的 API 密钥。
 * @param baseUrl - 服务基础地址。
 * @param timeoutMs - 单次请求超时，单位为毫秒。
 * @returns 密钥是否可继续复用；网络故障也返回真。
 */
export async function pingKey(
  apiKey: string,
  baseUrl: string,
  timeoutMs: number
): Promise<boolean> {
  // 仅 HTTP 401/403 表示密钥确定无效。
  // 网络故障不能视为无效密钥，避免重新创建账号并覆盖已有配置。

  try {
    const resp = await pingApiKey(baseUrl, apiKey, timeoutMs);
    return resp.status !== 401 && resp.status !== 403;
  } catch {
    warnOptional('key_validation_network');
    return true; // 状态未确认，不创建替代账号。
  }
}

/**
 * 存在调用方名称时同步平台身份及本地配置。
 * @param key - 访问服务的 API 密钥。
 * @param baseUrl - 服务基础地址。
 * @param agentCaller - 代理调用方名称。
 */
export async function maybeIdentify(
  key: string,
  baseUrl: string,
  agentCaller: string | undefined
): Promise<void> {
  // 复用密钥时同步显式声明的 agent_caller。

  if (!agentCaller) return;
  try {
    const resp = await identifyAgent(baseUrl, key, agentCaller);
    if (resp.ok) {
      try {
        const body = (await resp.json()) as { agent_caller?: string };
        if (fs.existsSync(CONFIG_FILE)) {
          const cfg = loadConfig();
          cfg.platform.agentCaller = body.agent_caller ?? agentCaller;
          saveConfig(cfg);
        }
      } catch {
        warnOptional('key');
      }
    }
  } catch {
    warnOptional('key');
  }
}

File: cli/python/src/memgo_cli/application/onboarding/key.py
"""已有密钥复用探测。"""

from __future__ import annotations

import httpx
from rich.console import Console

from memgo_cli.backend.auth import ping_key
from memgo_cli.runtime.warning import warn_optional

console = Console()
err_console = Console(stderr=True)


def _ping_key(api_key: str, base_url: str, timeout: float) -> bool:
    """探测现有密钥是否明确失效。

    仅 401/403 返回假；网络故障记录告警并保留既有密钥，避免误建账号。
    """
    try:
        resp = ping_key(base_url, api_key, timeout)
    except httpx.HTTPError as error:
        warn_optional("reuse_key_probe", error)
        return True  # 结果不确定时保留既有密钥
    return resp.status_code not in (401, 403)
